File: backend/lib/print/vertical_stroke_proportion.m
function results = compute_vertical_stroke_proportion(image_path, debug)
    % Check for debug argument
    if nargin < 2
        debug = false;
    end

    % Read the image
    img = imread(image_path);
    if isempty(img)
        error('Image file not found: %s', image_path);
    end

    % Convert to grayscale if needed
    if size(img, 3) == 3
        gray = rgb2gray(img);
    else
        gray = img;
    end

    % Apply thresholding using Otsu's method and invert binary image
    level = graythresh(gray);
    bw = imbinarize(gray, level);
    binary = ~bw;  % Invert: letters become white on black background

    % Find connected components (8-connected)
    cc = bwconncomp(binary, 8);
    stats = regionprops(cc, 'BoundingBox');

    % Initialize boxes and collect component heights
    boxes = [];
    heights = [];

    % Loop through each connected component
    for i = 1:numel(stats)
        bbox = stats(i).BoundingBox;  % [x, y, width, height]
        % Filter out small noise components (width and height > 5 pixels)
        if bbox(3) > 5 && bbox(4) > 5
            boxes = [boxes; bbox];
            heights = [heights; bbox(4)];
        end
    end

    results = struct();
    if ~isempty(heights)
        sortedHeights = sort(heights);
        median_height = median(sortedHeights);
        max_height = max(sortedHeights);
        ascender_ratio = max_height / median_height;
        results.median_height = median_height;
        results.max_height = max_height;
        results.ascender_ratio = ascender_ratio;
        fprintf('Ascender-to-xHeight ratio: %.2f\n', ascender_ratio);
    else
        warning('No connected components found with the given size threshold.');
    end

    % Debug visualization if requested
    if debug && ~isempty(boxes)
        % Prepare a copy of the image for drawing bounding boxes.
        % If the image is grayscale, convert it to RGB.
        if size(img, 3) == 1
            img_with_boxes = repmat(img, [1, 1, 3]);
        else
            img_with_boxes = img;
        end

        % Create a figure with three subplots
        figure('Name', 'Vertical Stroke Proportion Analysis', 'NumberTitle', 'off');

        % Subplot 1: Original Image
        subplot(1, 3, 1);
        imshow(img);
        title('Original Image');

        % Subplot 2: Binary Image
        subplot(1, 3, 2);
        imshow(binary);
        title('Binary Image');

        % Subplot 3: Bounding Boxes over Original Image
        subplot(1, 3, 3);
        imshow(img_with_boxes);
        hold on;
        for i = 1:size(boxes, 1)
            bbox = boxes(i, :);
            rectangle('Position', bbox, 'EdgeColor', 'g', 'LineWidth', 2);
        end
        title(sprintf('Bounding Boxes (Ascender Ratio: %.2f)', ascender_ratio));
        hold off;
    end
end

% Example usage:
image_path = '/Users/jameswong/PycharmProjects/NoteMercy_Extension/backend/atest/3.png';
results = compute_vertical_stroke_proportion(image_path, true);
disp(results);

File: backend/lib_py/italic/Vertical Stroke Proportion.py
import cv2
import matplotlib.pyplot as plt
import numpy as np
import base64
from io import BytesIO


class VerticalStrokeAnalyzer:
    def __init__(self, image_input, is_base64=True):
        """
        Initializes the VerticalStrokeAnalyzer with either a base64 encoded image or image path.

        Parameters:
            image_input (str): Either base64 encoded image string or image file path
            is_base64 (bool): If True, image_input is treated as base64 string, else as file path
        """
        if is_base64:
            # Decode base64 image
            img_data = base64.b64decode(image_input)
            nparr = np.frombuffer(img_data, np.uint8)
            self.img = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
            if self.img is None:
                raise ValueError("Error: Could not decode base64 image")
        else:
            # Read image from file path
            self.img = cv2.imread(image_input)
            if self.img is None:
                raise ValueError(f"Error: Could not read image at {image_input}")
        
        self.binary = None
        self.boxes = []
        self.metrics = {}

    def _preprocess_image(self):
        """Converts image to grayscale and applies thresholding to get a binary image."""
        # Convert image to grayscale
        gray = cv2.cvtColor(self.img, cv2.COLOR_BGR2GRAY)
        # Apply thresholding to get a binary image (inverted)
        _, self.binary = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)

    def _find_components(self):
        """Finds connected components (letters or parts of letters)."""
        # Find connected components
        num_labels, labels, stats, centroids = cv2.connectedComponentsWithStats(self.binary, connectivity=8)
        # Skip background (label 0) and filter out small components (noise)
        for i in range(1, num_labels):
            x = stats[i, cv2.CC_STAT_LEFT]
            y = stats[i, cv2.CC_STAT_TOP]
            w = stats[i, cv2.CC_STAT_WIDTH]
            h = stats[i, cv2.CC_STAT_HEIGHT]
            if h > 5 and w > 5:
                self.boxes.append((x, y, w, h))

    def _compute_metrics(self):
        """Computes the vertical stroke metrics including median height, max height, and ascender ratio."""
        # Extract the height of each component
        heights = [h for (x, y, w, h) in self.boxes]
        if heights:
            heights.sort()
            median_height = np.median(heights)  # Approximate x-height (middle zone)
            max_height = max(heights)  # Height of the tallest stroke (ascender or capital)
            ascender_ratio = max_height / median_height if median_height > 0 else 0

            self.metrics = {
                "median_height": float(median_height),
                "max_height": float(max_height),
                "ascender_ratio": float(ascender_ratio)
            }
        else:
            self.metrics = {
                "median_height": 0,
                "max_height": 0,
                "ascender_ratio": 0
            }

    def analyze(self, debug=False):
        """
        Analyzes the image to determine vertical stroke proportions.

        Parameters:
            debug (bool): If True, generates visualization plots.

        Returns:
            dict: Metrics including median height, max height, and ascender ratio,
                  plus visualization graphs if debug=True.
        """
        self._preprocess_image()
        self._find_components()
        self._compute_metrics()
        
        result = {
            'metrics': self.metrics,
            'graphs': []
        }
        
        # Generate visualization plots if debug mode is enabled
        if debug and self.boxes:
            # Draw bounding boxes on the original image
            img_with_boxes = self.img.copy()
            for (x, y, w, h) in self.boxes:
                cv2.rectangle(img_with_boxes, (x, y), (x + w, y + h), (0, 255, 0), 2)

            # Create a figure with subplots for visualization
            plt.figure(figsize=(15, 10))

            plt.subplot(131)
            plt.title("Original Image")
            plt.imshow(cv2.cvtColor(self.img, cv2.COLOR_BGR2RGB))
            plt.axis('off')

            plt.subplot(132)
            plt.title("Binary Image")
            plt.imshow(self.binary, cmap='gray')
            plt.axis('off')

            plt.subplot(133)
            plt.title(f"Bounding Boxes (Ascender Ratio: {self.metrics.get('ascender_ratio', 0):.2f})")
            plt.imshow(cv2.cvtColor(img_with_boxes, cv2.COLOR_BGR2RGB))
            plt.axis('off')

            plt.tight_layout()
            
            # Convert plot to base64
            buf = BytesIO()
            plt.savefig(buf, format='png', bbox_inches='tight')
            buf.seek(0)
            plot_base64 = base64.b64encode(buf.getvalue()).decode('utf-8')
            plt.close()
            
            result['graphs'].append(plot_base64)
            
        return result


# === Example usage ===
if __name__ == "__main__":
    # Example with file path
    image_path = "atest/3.png"
    analyzer = VerticalStrokeAnalyzer(image_path, is_base64=False)
    results = analyzer.analyze(debug=True)
    print(results['metrics'])
